CIPs/cip-draft_poprf.md

```diff
@@ -0,0 +1,474 @@
+used to construct an EIP-712 type.
+
+Other fields in the domain form the EIP-712 type. The EIP-712 type name should be equal to the
+interface name. Fields within the interface are sorted by name. Values SHOULD be mapped from their
+TypeScript types to the corresponding EIP-712 type. Specifically:
+
+* `string`  -> `string`
+* `number`  -> `int256`
+* `boolean` -> `bool`
+
+Optional values are encoded by interpreting it as EIP-712 type `Optional<T>(bool defined,T value)`,
+with `T` as the generic type name. When serialized, `value` MUST be set to the zero value of the
+type when `defined` is `false`. The zero value of a type is the value when all atomic types are set
+to zero, and dynamic types (`string` and `bytes`) are empty. Note that the zero hash of the encoded
+data of two distinct types may be distinct (i.e. it is not equal to the empty hash).
+
+As an example, the following instance of a `LinearBackoffDomain` is hashed:
+```typescript
+const domain: LinearBackoffDomain = {
+  name: "ODIS Linear Backoff Domain"
+  version: 1
+  cap: 10
+  salt: "saltvalue"
+}
+```
+
+<!-- TODO(victor) Verify that there are no mistakes in this example. -->
+```text
+domainSeparator = keccak256(
+  keccak256("EIP712Domain(string name,string version)") ||
+  keccak256("ODIS Linear Backoff Domain") || keccak256("1")
+)
+typeHash = keccak256(
+  "LinearBackoffDomain(int256 cap, Optional<int256> refresh, Optional<string> salt)" ||
+  "Optional<int256>(bool defined,int256 value)" ||
+  "Optional<string>(bool defined,string value)"
+)
+structHash = keccak256(
+  typeHash ||
+  0x000000000000000000000000000000000000000000000000000000000000000A ||
+  keccak256(
+    keccak256("Optional<int256>(bool defined,int256 value)") ||
+    0x0000000000000000000000000000000000000000000000000000000000000000 ||
+    0x0000000000000000000000000000000000000000000000000000000000000000
+  ) ||
+  keccak256(
+    keccak256("Optional<string>(bool defined,string value)") ||
+    0x0000000000000000000000000000000000000000000000000000000000000001 ||
+    keccak256("saltvalue")
+  )
+)
+encoded = 0x1901 || domainSeparator || structHash
+```
+
+### Signatures
+
+Signatures are a useful primitive for authenticated rate limiting (e.g. allowing a higher rate limit
+to the owners of verified accounts on Celo). In order to facilitate this, the entire request struct
+should be deterministically serializeable, specifically using EIP-712.
+
+Domains that make use of signatures for authentication of requests (e.g. the `PepperedDomain`
+example below) should include a field in the domain options for specification of a signature.
+Additional fields may also be added, such as a nonce to prevent replaying requests.
+
+Signatures should be verified over the entire request, excluding the signature itself. Specifically,
+the message against which a signature is verified should be obtained by first removing the signature
```

Comment:
Does it matter if "removing" the signature means removing the entire field or just setting it to its zero value?

Reply:
I was thinking the field should be removed entirely from the struct, but on second thought will change it to set it to zero instead.

Reply:
Oh and to answer your direct question, both work as long as implementations are consistent.